### ISSUE TITLE
Regex capture groups in nonterminal patterns

### ISSUE THREAD

**Issue by martin-danhier** [lalrpop/lalrpop#575]
Hello everyone,

While trying to write a simple LALRPOP grammar, I ran into a recurrent issue:

For example, take the following grammar that defines a simple CharLiteral nonterminal. Here, a char is a single character between single quotes, like `'a'`, `'d'`,  or `' '`. It doesn't support escape characters yet.

```rust
CharLiteral: char = {
    "'" <value: r"[^\s\\']| "> "'" => value.chars().nth(0).unwrap()
}
```

The problem with this nonterminal is that LALRPOP skips spaces between each token of the pattern. For example, `' c '` would be considered a valid CharLiteral, which is not desired.

A way of solving this problem would be to use a Regex capture group:

```rust
CharLiteral: char = {
    <value: r"'([^\s\\']| )'"> => ... // Use the value of the first capture group
}
```

However, LALRPOP doesn't apparently support capture groups. Here, in the action code, `value` contains a `str`. So, if we want to get the value of the first capture group, we need to manually match the regex a second time in the action code: 

```rust
CharLiteral: char = {
    <value: r"'([^\s\\']| )'"> => {
        // Create regex
        let re = Regex::new(r"'([^\s\\']| )'").unwrap();
        let c = re.captures(value).unwrap();
        // Use result from first capturing group
        c.get(1).unwrap().as_str().chars().nth(0).unwrap()
    }
}
```

This works, but I don't think it is a good solution :
- You have to write the regex two times, which is error prone when you modify it later (you can forget to update the other one)
- The same string is analyzed two times, which I think is less efficient
- It is needed often, so when you have 10 of these it bloats the grammar

I have two ideas to help solve this problem in a better way :

### 1. Support capture groups

Currently, the syntax `<>` returns the whole matched string. 
When a regex is used in a pattern, a similar syntax, let's say `<c>` , could return the "captures" object instead of the matched string.

```rust
CharLiteral: char = {
    <r"'([^\s\\']| )'"> => {
        // Use result from first capturing group
        <c>.get(1).unwrap().as_str().chars().nth(0).unwrap()
    }
}
```

More, using a number between the brackets could be used to direclty access a specific group.
For example, `<1>` would return an Option containing either the match of a specific group or None if there isn't (it would be the equivalent of `<c>.get(1)`.

With that syntax, the nonterminal will look like this:

```rust
CharLiteral: char = {
    <r"'([^\s\\']| )'"> =>  <1>.unwrap().as_str().chars().nth(0).unwrap()
}
```

Which is simpler and cleaner in my opinion.

### 2. Specific syntax to disable space skipping between tokens

In a pattern, a character, like a tilde, could be used between tokens to specify that there must not be any space between them.

```rust
CharLiteral: char = {
    "'" ~ <value: r"[^\s\\']| "> ~ "'" => value.chars().nth(0).unwrap()
}
```

This would match `'c'` but not `' c '`.

However, this solution would only solve similar space-related issues, where the capture groups would be useful in other cases as well.

What do you think about this ?
If you think that it is a good idea, I'm willing to dive into the LALRPOP code to find a way of implementing it.

PS: Happy new year !

**Comment by zoombinis** [comment]
I seem to be correctly parsing the `c` in `' c '` with:

```rust
<value:("'" <r"[^\s\\']| "> "'")> => value.to_string()
```

@martin-danhier Does this get you any closer?

**Comment by martin-danhier** [comment]
Thanks for your reply @zoombinis !

The problem here isn't about managing to parse the `c` in `' c '`, because it works using the simplest form of the rule:

```rust
"'" <value: r"[^\s\\']| "> "'" => value.chars().nth(0).unwrap()
```

The main problem that I had with that is that I wanted to forbid the use of spaces inside the quotes:

- `'c'` -> allowed
- `' c'`, `'c ', ``' c '`, ... -> error

With the rule mentioned above, all of these parse successfully because LALRPOP ignores spaces between each token. That is why it would be useful to have either a way to locally disable skipping between specific tokens or a way to support regex capture groups.

But I'd be glad to hear if there is already an official way to achieve the same result.

**Comment by zoombinis** [comment]
oh, I see what you mean. I'm new to lalrpop, so I can't tell if this is a bug or not but actually lalrpop _does_ currently (`0.19.6`) recognize space, though it is not something I'm entirely understanding correctly.

As a simple example, I am capturing `'c'` as "good" and `' c '` as "bad" with:

```rust
Attr: String = {
  r"'[a-z]'" => "good".to_string(),
  r"' [a-z] '" => "bad".to_string(),
}
```

**Comment by FranklinChen** [comment]
I agree it would be handy to support capture groups from regexes, because right now I am also re-matching when delving into token substructure.

**Comment by cdecompilador** [comment]
Is somebody working on this issue already?
